Don't flow self contained command-line option to referenced projects

dsplaisted opened the pull request:
Fixes dotnet/sdk#21677

### Context

In .NET 6, if you run a command such as `dotnet build -r win-x64`, you will get the following error:

> C:\Program Files\dotnet\sdk\6.0.100-rtm.21474.29\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(1105,5): warning NETSDK1179: One of '--self-contained' or '--no-self-contained' options are required when '--runtime' is used. [C:\git\repro\ConsoleTest\ConsoleTest.csproj]

However, if you follow the suggestion and change it to `dotnet build -r win-x64 --self-contained`, then you will likely get an error such as the following if there are any referenced projects:

> C:\Program Files\dotnet\sdk\6.0.100-rc.2.21501.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(150,5): error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false. [C:\Users\brady\source\TodoApp\TodoApp.API\TodoApp.API.csproj]

This happens because

- Global properties (ie those specified on the command line) flow to referenced projects by default
- We normally stop the `RuntimeIdentifier` from flowing to referenced projects
- We don't currently stop the `SelfContained` property from flowing to referenced projects
- So the referenced project gets built with `SelfContained` set to true but no `RuntimeIdentifier`, which is an error

### Changes Made

If `--self-contained` is specified on the command-line, and the `RuntimeIdentifier` is blocked from flowing to referenced projects, then we also block the `SelfContained` property from flowing to referenced projects.

Note that this treats `--self-contained` differently from `/p:SelfContained=true`, as there is at least one project that relies on `/p:SelfContained=true` flowing across project references, and they set `RuntimeIdentifier` in their .props files if `SelfContained` is true.

### Testing

I will add automated tests to the SDK repo for this scenario, and test locally with those tests and the changes in this PR before merging.

### Notes

FYI @richlander @sfoslund @marcpopMSFT, I think we should take this for 17.0 GA

BenVillalobos commented:
If `_CommandLineDefinedSelfContained` is the expected escape hatch,  why mark it with an underscore? 

I'm a little confused at why there's a distinction between --self-contained and /p:SelfContained=true, but I won't block on it.

dsplaisted commented:
More background than you probably want to dig into is available here: https://github.com/dotnet/designs/blob/main/accepted/2021/architecture-targeting.md

We wouldn't expect someone to define `_CommandLineDefinedSelfContained` themselves, the CLI automatically sets it if you use the `--self-contained` or `--no-self-contained` options: https://github.com/dotnet/sdk/blob/f72371bb4e6b5d3394ec2c261cf8c43a6697b815/src/Cli/dotnet/CommonOptions.cs#L115-L125

So the escape hatch is basically to use `/p:SelfContained=true` instead of `--self-contained`.

@Forgind We shouldn't merge this yet until I've had a chance to test it.  I plan on writing automated tests in dotnet/sdk, but we won't be able to enable them in CI until the fix merges and flows.  But I should still run them manually first.

dsplaisted commented:
I've done some more work and I think I have a pretty good idea of how we should solve this.  It involves changes to MSBuild (now included in this PR) as well as to the SDK, in dotnet/sdk#21986.

I'd like some feedback on whether that's the right direction.  It will be a breaking change (see the breaking change document in the SDK PR), so I think we *probably* would want to target these fixes for 6.0.200 rather than 6.0.100 GA.

dsplaisted commented:
@rainersigwald I've retargeted this to vs17.1, and I think it (together with https://github.com/dotnet/sdk/pull/21986) is ready for review again

rainersigwald commented:
Our primary 17.1 branch is `main` so I retargeted for you again. Reviewing.

beatcracker commented:
I'm looking into this due to dotnet/sdk/issues/10566, and I've found that just checking for `_CommandLineDefinedRuntimeIdentifier` works for my simple case (no `rid` in the project, `self-contained` / `rid` are supplied via the commandline):

```xml
<IsRidAgnostic Condition=" '$(RuntimeIdentifier)' == '' and '$(RuntimeIdentifiers)' == '' and $(_CommandLineDefinedRuntimeIdentifier) == '' ">true</IsRidAgnostic>
```

Obviously, I've just skimmed the target source and I have no deep knowledge, but I hope this helps.

Forgind commented:
@dsplaisted, are you going to be able to get around to this soon? Or should we close it for now or mark it as a draft?

dsplaisted commented:
@Forgind Is there a way to convert a pull request back to a draft PR?

dsplaisted commented:
@rainersigwald @dotnet/msbuild I've updated and rebased this PR, and I believe it is now ready for review again.  I switched back to using `IsRidAgnostic` instead of creating `AcceptsRuntimeIdentifier`, which should help make this more compatible.

Forgind commented on src/Tasks/Microsoft.Common.CurrentVersion.targets:
```diff
@@ -1863,9 +1863,12 @@ Copyright (C) Microsoft Corporation. All rights reserved.
+           unless the project is expecting those properties to flow.
+           We include SelfContained together with RuntimeIdentifier, because otherwise a referenced project could try to build with
+           SelfContained set to true but no RuntimeIdentifier set, which causes an error (for Exe projects). -->
+      <AnnotatedProjects Condition="'@(AnnotatedProjects)' == '%(Identity)' and '%(AnnotatedProjects.IsRidAgnostic)' != 'false'">
```
When is '@(AnnotatedProjects)' != '%(Identity)'?

dsplaisted replied:
I think you have to do this when updating metadata this way, otherwise each update will apply to all the items, or something like that.  @rainersigwald probably knows the details.  Anyway my changes don't change that pattern here, they just change the other conditions.

Forgind commented on src/Tasks/Microsoft.Common.CrossTargeting.targets:
```diff
@@ -42,9 +42,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
-        <!-- indicate to caller that project is RID agnostic so that a global property RuntimeIdentifier value can be removed -->
-        <IsRidAgnostic>false</IsRidAgnostic>
-        <IsRidAgnostic Condition=" '$(RuntimeIdentifier)' == '' and '$(RuntimeIdentifiers)' == '' ">true</IsRidAgnostic>
+        <IsRidAgnostic>@(_TargetFrameworkInfo->'%(IsRidAgnostic)')</IsRidAgnostic>
```
Should GetTargetFrameworks depend on GetTargetFrameworksWithPlatformForSingleSingleTargetFramework explicitly?

dsplaisted replied:
That's what it does for single-targeted projects: https://github.com/dotnet/msbuild/blob/5d102ae37163fc1f22610bb433c3ab9a2fff91f0/src/Tasks/Microsoft.Common.CurrentVersion.targets#L1889-L1890

For multi-targeted projects it has to do inner builds for each target framework, which the `GetTargetFrameworksWithPlatformFromInnerBuilds` target handles.

Forgind commented on src/Tasks/Microsoft.Common.CurrentVersion.targets:
```diff
@@ -1779,7 +1779,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         Properties="%(_MSBuildProjectReferenceExistent.SetConfiguration); %(_MSBuildProjectReferenceExistent.SetPlatform)"
         ContinueOnError="!$(BuildingProject)"
-        RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove);TargetFramework;RuntimeIdentifier$(_GlobalPropertiesToRemoveFromProjectReferences)"
+        RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove);TargetFramework;RuntimeIdentifier;SelfContained;$(_GlobalPropertiesToRemoveFromProjectReferences)"
```
It was previously RuntimeIdentifier$(_GlobalPropertiesToRemoveFromProjectReferences) (no semicolon)
That looks like a bug to me, but I just wanted to make sure that's intentional 🙂

dsplaisted replied:
Good point.  It does seem like it might have been a bug before.  Now there may be extra semicolons, but that should be OK, right?  The MSBuild task should just ignore the extra ones I think.

Forgind replied:
Yes

Forgind commented on src/Tasks/Microsoft.Common.CurrentVersion.targets:
```diff
@@ -1941,6 +1942,16 @@ Copyright (C) Microsoft Corporation. All rights reserved.
+             Generally this value will come from the IsRidAgnostic property set by the .NET SDK.  If that's not set, then the fallback logic here will be that the project
+             is RID agnostic if it doesn't have RuntimeIdentifier or RuntimeIdentifiers properties set. -->
+        <IsRidAgnostic>$(IsRidAgnostic)</IsRidAgnostic>
+        <IsRidAgnostic Condition=" '$(IsRidAgnostic)' == '' and '$(RuntimeIdentifier)' == '' and '$(RuntimeIdentifiers)' == '' ">false</IsRidAgnostic>
```
Should this be true and the next one false?

joeloff replied:
+1, it looks like it will set `IsRidAgnostic` to false when there are no runtime identifiers

dsplaisted replied:
Yes, thanks for catching this.

Forgind commented on src/Tasks/Microsoft.Common.CurrentVersion.targets:
```diff
@@ -1941,6 +1942,16 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         <TargetPlatformMonikers>$(TargetPlatformMoniker)</TargetPlatformMonikers>
         <TargetPlatformMonikers Condition="'$(TargetPlatformMoniker)' == ''">None</TargetPlatformMonikers>
         <AdditionalPropertiesFromProject>$(_AdditionalTargetFrameworkInfoProperties)</AdditionalPropertiesFromProject>
+
```
Does _TargetFrameworkInfo ever have more than just $(TargetFramework) included in it?

dsplaisted replied:
Yes, it does for multi-targeted projects, which will run this target for each target framework and combine the results.

joeloff commented on src/Tasks/Microsoft.Common.CurrentVersion.targets:
```diff
@@ -1795,7 +1795,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         BuildInParallel="$(BuildInParallel)"
         ContinueOnError="!$(BuildingProject)"
-        RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove);TargetFramework;RuntimeIdentifier;Platform;Configuration$(_GlobalPropertiesToRemoveFromProjectReferences)"
+        RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove);TargetFramework;RuntimeIdentifier;SelfContained;Platform;Configuration$(_GlobalPropertiesToRemoveFromProjectReferences)"
```
Do we need a `;` between `Configuration` and `$(_GlobalPropertiesToRemoveFromProjectReferences)`?

Forgind replied:
Looking at all usages of _GlobalPropertiesToRemoveFromProjectReferences in MSBuild, it seems that it is always expected to have a `;` in front of it. That said, I think adding a ; is better anyway.

Forgind commented on src/Tasks/SetRidAgnosticValueForProjects.cs:
```diff
@@ -0,0 +1,65 @@
+                .Select(p =>
+                {
+                    var hasSingleTargetFrameworkString = p.GetMetadata("HasSingleTargetFramework");
+                    if (!ConversionUtilities.ValidBooleanFalse(hasSingleTargetFrameworkString))
```
Why did you choose !false rather than true? Could it be null here? If so, this seems like wrong behavior.

dsplaisted replied:
If the value is not set I want this to be a no-op.  That seems the safest.

For projects using the .NET SDK, this should always be set.  For other types of projects it might not be.

Forgind commented on src/Tasks/SetRidAgnosticValueForProjects.cs:
```diff
@@ -0,0 +1,65 @@
+                    var hasSingleTargetFrameworkString = p.GetMetadata("HasSingleTargetFramework");
+                    if (!ConversionUtilities.ValidBooleanFalse(hasSingleTargetFrameworkString))
+                    {
+                        // No change to item, it should already have a single-valued IsRidAgnostic value
```
Should we have a Debug.Assert here or something?

dsplaisted replied:
`Debug.Assert` gets compiled out in release builds, so that would be pretty unlikely to catch anything anyway, wouldn't it be?

Forgind replied:
It is, but if something goes wrong, we'd probably try to debug with a debug build, so I thought it might be helpful. Not a big deal if you don't want to add it.

Forgind commented on src/Tasks/SetRidAgnosticValueForProjects.cs:
```diff
@@ -0,0 +1,65 @@
+                    }
+                    var updatedItem = new TaskItem(p);
+
+                    var nearestTargetFramework = p.GetMetadata("NearestTargetFramework");
```
Are projects actually expected to have all of this information? I feel like this relies on projects carrying around a lot of (generally unimportant) baggage. Not saying that doesn't happen, but it's a bit surprising to me.

dsplaisted replied:
What information are you referring to?  These are part of the [Project Reference Protocol](https://github.com/dotnet/msbuild/blob/main/documentation/ProjectReference-Protocol.md#targets-required-to-be-referenceable), or in the case of `NearestTargetFramework`, set by the `GetReferenceNearestTargetFrameworkTask` task which runs just before this one.

Forgind replied:
I wasn't aware of NearestTargetFramework, but that makes sense. I was also surprisedby IsRidAgnostic being an array of IsRidAgnostic values with one per target framework. I'd expected (and the Project Reference Protocol seems to say) that it's just a boolean. It seems like you're changing the type in this task from boolean[] to boolean?

Forgind commented on src/Tasks/Microsoft.Common.CurrentVersion.targets:
```diff
@@ -1850,6 +1850,20 @@ Copyright (C) Microsoft Corporation. All rights reserved.
+    <ItemGroup>
+      <AnnotatedProjects Remove="@(AnnotatedProjects)" />
+      <AnnotatedProjects Include="@(UpdatedAnnotatedProjects)" />
+      <UpdatedAnnotatedProjects Remove="@(UpdatedAnnotatedProjects)" />
```
Why is this line necessary?

dsplaisted replied:
It just clears it out so that it's not taking up memory anymore.  It's a pretty clunky pattern for when you want to have a task that updates item metadata.

Forgind reviewed (approved):
I'm least confident about the IsRidAgnostic array, and I haven't had a chance to manually test this myself, but I don't know of anything broken in here, you said you wanted it in this week, and I have something in five minutes 🙂
